### PR TITLE
feat(community): Neo4j chat message history

### PR DESCRIPTION
**Pull request by BernardFaucher**
## What? Why?
This PR adds the Neo4jChatMessageHistory class to `@langchain/community`, in furtherance of reaching functional parity with the Python version of the package. PTAL 👀

## Tests?
Yes! A new integration test is added to the corresponding `tests` folder, covering all the new functionality in well-scoped `describe/it` blocks, including set-up and tear-down. To facilitate testing, the neo4j service has also been added to `test-int-deps-docker-compose.yml`, pinned to the latest image. To start the test container, run: `docker-compose -f test-int-deps-docker-compose.yml up -d` from the project root, then run the test with: `yarn run test:single src/stores/tests/neo4j.int.test.ts` from the `libs/langchain-community` workspace. Alternatively, you can test against a neo4j server running locally with Neo4j Desktop, or in the cloud using https://sandbox.neo4j.com/ - all three options are working.

## Demo?
Yes! Here's a gif of the test successfully passing:
neo4jChatMessageHistoryTest

And here's a png of what a chat message history persisted with `Neo4jChatMessageHistory` looks like in neo4j:
persistedChatHistory

## Considerations? 
Where sensible, the defaults and conventions of the python code are maintained. Any divergences are configurable, and represent a holistic view of the js/node ecosystem and coding conventions. Raw Cypher passed to the neo4j driver via the new class is tightened up and idiomatic, without hurting query functionality, and uses the params object over string interpolation as much as possible as a security consideration.

## Appreciation?
Twitter is not the place to be, but I'd love a friendly shout out on Mastodon at [@bernardffaucher@ioc.exchange](https://ioc.exchange/@bernardffaucher) or mention [on my LinkedIn](https://www.linkedin.com/in/bernardffaucher/) 🙏

### CODE REVIEW

File: libs/langchain-community/src/stores/message/neo4j.ts [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+const defaultConfig = {
+  sessionNodeLabel: "ChatSession",
+  messageNodeLabel: "ChatMessage",
+  windowSize: 3,
+}
+
+export class Neo4jChatMessageHistory extends BaseListChatMessageHistory {
+
+  lc_namespace: string[] = ["langchain", "stores", "message", "neo4j"];
```
Comment (jacoblee93) [inline libs/langchain-community/src/stores/message/neo4j.ts:28]: Make sure you run `yarn format`!

File: libs/langchain-community/src/stores/message/neo4j.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+  sessionId: string | number;
+  sessionNodeLabel: string;
+  messageNodeLabel: string;
+  windowSize: number;
+
+  #driver: Driver;
```
Comment (jacoblee93) [inline libs/langchain-community/src/stores/message/neo4j.ts:34]: We generally use the `private` modifier for private properties as well

File: libs/langchain-community/src/stores/message/neo4j.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    super()
+
+    this.sessionId = sessionId;
+    this.sessionNodeLabel = sessionNodeLabel;
+    this.messageNodeLabel = messageNodeLabel;
+    this.windowSize = windowSize;
+
+    if (url && username && password) {
+      try {
+        this.#driver = neo4j.driver(url, auth.basic(username, password));
```
Comment (jacoblee93) [inline libs/langchain-community/src/stores/message/neo4j.ts:54]: This is sync?

**Review by jacoblee93 — COMMENTED** [review]
Looks ok overall, will tag in @tomasonjo for a look too though!

**Comment by tomasonjo** [comment]
No, no need for arbitrary queries. Just change session run to driver
execute query

V pet., 13. dec. 2024, 17:02 je oseba Bernard F Faucher <
***@***.***> napisala:

**Review by jacoblee93 — APPROVED** [review]
Thank you!

**Comment by jacoblee93** [comment]
`Neo4jError` seems to take an additional argument for code, just switched to native error

**Pull request merged**